### ISSUE TITLE
Non–late-escaping XSS sniff

### ISSUE THREAD

**Issue by JDGrimes** [WordPress-Coding-Standards/WordPress-Coding-Standards#845]
The XSS sniff that is currently a part of the standards is designed to enforce a philosophy of late escaping—everything should be escaped immediately before output. However, strict adherence to this is currently difficult, especially for themes, because of some patterns used in WordPress core. (Many WordPress functions return a string, rather than echoing it directly, and do not always escape everything that is incorporated into the output.) There are also times, like when reviewing code, where one is mostly interested in detecting low-hanging fruit, not forcing a late-escaping policy on another developer.

To remedy this, a new sniff may need to be introduced, though it is possible that the existing sniff could also be modified to have a less strict "mode". The idea would be to only flag unescaped output that incorporated known-to-be-dangerous values, like `$_POST`, `$_GET`, etc.

Related: #748, https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/844#issuecomment-279714754

**Comment by westonruter** [comment]
What about using `error` for the low-hanging fruit and `warning` for anything in strict mode?

**Comment by grappler** [comment]
> What about using `error` for the low-hanging fruit and `warning` for anything in strict mode?

I think it would depend on the ruleset.

So we will be having a list of three different types of functions.

## Insecure function
- `get_theme_mod`
- `get_options`
- `$_GET`
- `$_POST`

## Mostly secure functions
- `get_the_category_list`

## Totally secure functions
> Any non-pluggable function that returns a known limited set of safe static values not directly based on the input arguments
- `category_description`

My last question would be for the totally secure function the data would need to be escaped after the filter and not before?

**Comment by JDGrimes** [comment]
> What about using error for the low-hanging fruit and warning for anything in strict mode?

That is another possible approach. It really doesn't matter to me, it depends on the different applications that people have for this. As @grappler, it might depend on the ruleset.

>So we will be having a list of three different types of functions.

Yes, probably something like that.

> My last question would be for the totally secure function the data would need to be escaped after the filter and not before?

Yes.

**Comment by dingo-d** [comment]
I've been going through the list as well, and so far I've found 2 deprecated functions in the `$autoEscapedFunctions` list.

## Deprecated functions

 ~~~
comments_popup_script
comments_rss_link
~~~

`bloginfo()` is safe as it echoes

~~~
echo get_bloginfo( $show, 'display' );
~~~

The `display` filter is safe to use (as commented [here](http://wordpress.stackexchange.com/a/80449/58895)).

So far on my list of `$autoEscapedFunctions` are the following functions

~~~
allowed_tags
bloginfo
body_class
calendar_week_mod
category_description
checked
comment_author_IP
comment_form
comments_link
count
delete_get_calendar_cache
disabled
do_shortcode
~~~

I'm still going through the list, will update as I check more.

**Comment by grappler** [comment]
`get_bloginfo()`  is not valid for `$autoEscapedFunctions` as the data can be modified via a filter that is why I have added to the `Mostly secure functions`.

Some of the items on your list conflict with the items on my list. https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/845#issuecomment-279834686

**Comment by dingo-d** [comment]
I just checked if they are safe, I didn't check for 'mostly safe', that's why I have these still in the allow :)

And `get_bloginfo` isn't on the `$autoEscapedFunctions` list, so we're good there I think.

**Comment by jrfnl** [comment]
@dingo-d Regarding deprecated functions being on the lists: IMHO they should be (correctly detected) for the purposes of these lists.
There is a separate sniff to alert about the use of deprecated functions.

**Comment by dingo-d** [comment]
@jrfnl So the fact that they are on the list doesn't matter? Seems like a bit of an unnecessary clutter. If we are sorting these out, and have a separate sniff for deprecated functions, I think there's no need for them to be here.

**Comment by jrfnl** [comment]
@dingo-d I do not agree. These are two separate issues and should be sniffed for independently of each other. Someone might have either one of these sniffs disabled or use a ruleset which doesn't include one of them (`VIP`), the other sniff should still function correctly in that case. So deprecated functions should stay in the list.

**Comment by grappler** [comment]
@JDGrimes @jrfnl I have gone through the current list of [`$autoEscapedFunctions`](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/WordPress/Sniff.php#L85-L229) and have seperated them into different sections. https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/845#issuecomment-279834686

I have just used the functions that we currently have listed. There are some functions missing but I think it would be easiest to add them once we have the changes.

There are a few functions that need to be double checked and a few that need a decision what to do with them.

**Comment by JDGrimes** [comment]
>I have gone through the current list of $autoEscapedFunctions and have seperated them into different sections. #845 (comment)

Looks good. (I haven't double checked any of them yet, though.)

>I have just used the functions that we currently have listed. There are some functions missing but I think it would be easiest to add them once we have the changes.

+1.

>There are a few functions that need to be double checked and a few that need a decision what to do with them.


>  - `comments_popup_script` is deprecated and does nothing.

Yeah, might as well remove it.

>  - `delete_get_calendar_cache` does not have a return value so it should never be echoed

How did it even get on the list? 😄 (Just a rhetorical question, let's remove it.)

>  - `do_shortcode_tag` is a private function and in core it is not echoed

I don't see why that would ever need to be echoed directly. It's not what it is designed for, and as you say, it is private, so I see no reason to keep in on the list.

> My last question would be for the totally secure function the data would need to be escaped after the filter and not before?

Yes. The only exception would be for functions that are specifically escaping functions. For example, `esc_html()` has a filter, but it is obvious to anything that is using that filter that they need to be escaping whatever they return. For these functions with filters that aren't specifically for applying escaping to a value, they should not have any filter after they escape the data. If they do, then they don't belong on the totally secure function list.
